src/Xamarin.Android.Build.Tasks/Utilities/CecilExtensions.cs

```diff
@@ -0,0 +1,186 @@
+	public static IEnumerable<RegisterAttribute> GetMethodRegistrationAttributes (Mono.Cecil.ICustomAttributeProvider p)
+	{
+		foreach (var a in CecilExtensions.GetAttributes<RegisterAttribute> (p, a => CecilExtensions.ToRegisterAttribute (a))) {
+			yield return a;
+		}
+		foreach (var c in p.GetCustomAttributes ("Java.Interop.JniConstructorSignatureAttribute")) {
+			var r = RegisterFromJniConstructorSignatureAttribute (c);
+			if (r == null) {
+				continue;
+			}
+			yield return r;
+		}
+		foreach (var c in p.GetCustomAttributes ("Java.Interop.JniMethodSignatureAttribute")) {
+			var r = RegisterFromJniMethodSignatureAttribute (c);
+			if (r == null) {
+				continue;
+			}
+			yield return r;
+		}
+	}
+
+	internal static RegisterAttribute? RegisterFromJniMethodSignatureAttribute (CustomAttribute attr)
+	{
+		// attr.Resolve ();
+		RegisterAttribute? r = null;
+		if (attr.ConstructorArguments.Count == 2)
+			r = new RegisterAttribute ((string) attr.ConstructorArguments [0].Value,
+				(string) attr.ConstructorArguments [1].Value,
+				"",
+				attr);
+		return r;
+	}
+
+	internal static RegisterAttribute? RegisterFromJniConstructorSignatureAttribute (CustomAttribute attr)
+	{
+		// attr.Resolve ();
+		RegisterAttribute? r = null;
+		if (attr.ConstructorArguments.Count == 1)
+			r = new RegisterAttribute (
+				name: ".ctor",
+				signature: (string) attr.ConstructorArguments [0].Value,
+				connector: "",
+				originAttribute: attr);
+		return r;
+	}
+
+	// Keep in sync w/ GetMethodRegistrationAttributes()
+	public static bool HasMethodRegistrationAttributes (Mono.Cecil.ICustomAttributeProvider p)
+	{
+		foreach (CustomAttribute custom_attribute in p.CustomAttributes) {
+			var customAttrType = custom_attribute.Constructor.DeclaringType.FullName;
+			foreach (var t in MethodRegistrationAttributes) {
+				if (customAttrType == t)
+					return true;
+			}
+		}
+		return false;
+	}
+
+	public static IEnumerable<TAttribute> GetAttributes<TAttribute> (Mono.Cecil.ICustomAttributeProvider p, Func<CustomAttribute, TAttribute?> selector)
+			where TAttribute : class
+	{
+		return GetAttributes (p, typeof (TAttribute).FullName!, selector);
+	}
+
+	public static IEnumerable<TAttribute> GetAttributes<TAttribute> (Mono.Cecil.ICustomAttributeProvider p, string attributeName, Func<CustomAttribute, TAttribute?> selector)
+			where TAttribute : class
+	{
+		return p.GetCustomAttributes (attributeName)
+			.Select (selector)
+			.Where (v => v != null)
+			.Select (v => v!);
+	}
+
+	internal static IEnumerable<RegisterAttribute> GetTypeRegistrationAttributes (Mono.Cecil.ICustomAttributeProvider p)
+	{
+		foreach (var a in CecilExtensions.GetAttributes<RegisterAttribute> (p, a => CecilExtensions.ToRegisterAttribute (a))) {
+			yield return a;
+		}
+		foreach (var c in p.GetCustomAttributes ("Java.Interop.JniTypeSignatureAttribute")) {
+			var r = RegisterFromJniTypeSignatureAttribute (c);
+			if (r == null) {
+				continue;
+			}
+			yield return r;
+		}
```

Comment:
Does it matter the ordering these are `yield returned`?

It seems like a single `foreach` loop would allocate a lot less and look for the two attributes by name. But I don't know if the ordering matters.

Reply:
All of these functions were copy/pasted from the `internal` [CecilExtensions.cs](https://github.com/dotnet/java-interop/blob/main/src/Java.Interop.Tools.JavaCallableWrappers/Java.Interop.Tools.JavaCallableWrappers/CecilExtensions.cs) in `Java.Interop.Tools.JavaCallableWrappers` because we need them outside the JCW process.  Once we figure out where everything ends up, we will probably need to determine where functions like this need to live if they are still needed by multiple processes.